File: Partition_label.py
"""
Time Complexity : O(2n)- n for making a dictionary and another n for traversing though and calculating
Space Complexity : O(1)- as result is just auxillary space and dictionary can have maximum 26 chars
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
Your code here along with comments explaining your approach:
Here, we will first build a dictionary with last occurances of all the characters. We can then start from the 
very first character and check when does it occur last. We will traverse till the last position and if 
in between, any other character's last index is beyond pur last index, we will update our end pointer.
When we finally reach end, we will update our result and start traversing with the new character from 
a new start.
"""


class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        dic = {}
        result = []
        
        for i in range(len(S)):
            dic[S[i]] = i
            
        start_idx = 0
        end_idx = 0
        for i in range(len(S)):
            end_idx = max(end_idx, dic[S[i]])
            if i == end_idx:
                result.append(i - start_idx + 1)
                start_idx = i + 1
        return result

File: Partition-Label.java
//Time Complexity: o(n)
//Space Complexity: o(n)
//Expln: Store the last index of each element in the hashmap. Iterate over the hashmap
// keep updating the end by max from the  hashmap for every Character, whenever the index from array 
// meets the end that means thats the last position for the occurence of the char. Store the value of last position
// update start to next char and do this till end of the string.
class Solution {
    public List<Integer> partitionLabels(String S) {
        Map<Character, Integer> map = new HashMap<>();
        List<Integer> res = new ArrayList<>();
        if(S.length() == 0) return res;
        
        for(int i = 0; i < S.length(); i++)
        {
            char c = S.charAt(i);
            map.put(c, i);
        }
        
        int start = 0; int end = 0;
        for(int i = 0; i < S.length(); i++)
        {
            char ch = S.charAt(i);
            end = Math.max(end, map.get(ch));
            if(i == end)
            {
                res.add(end-start+1);
                start  = end+1;
            }
        }
        return res;
    }
}